### PR TITLE
Update to use new upstream unified keyboard branch

### PR DESCRIPTION
**Pull request by Kethku**
## What kind of change does this PR introduce?
- Fix

## Did this PR introduce a breaking change? 
_A breaking change includes anything that breaks backwards compatibility either at compile or run time._
This may break ime support. Not sure, this needs to be tested.

### CODE REVIEW

**Comment by Kethku** [comment]
Note: the ime support seems way better in the new upstream branch, so at some point after this merges, we should pipe the cursor positions to the ime system. Should resolve some long standing bugs.

We also get precommit texts now for ime, so it should also be possible to render in progress texts next under the cursor to fully round out our ime support.

**Comment by shaunsingh** [comment]
image

Small fix to the our glutin fork to build properly on macOS, we can/should upstream it though

**Comment by shaunsingh** [comment]
`Cargo.toml` also still puts `winit` at the old `new-keyboard-all` branch, not sure if that should be updated to v3

**Comment by fredizzimo** [comment]
I think we need to prioritize getting this fixed and merged. `wgpu` will need a more recent winit. And so will my changes to the update loop, so that I can read the display refresh rate.

As far as I see, it's failing because of glutin, we should probably update that to the latest as well. We will need some changes to the context creation, but the `glutin-winit` makes it look almost the same as before.

**Comment by fredizzimo** [comment]
@Kethku, @shaunsingh. I updated to the latest Glutin and cherry-picked the relevant changes from here https://github.com/neovide/neovide/pull/1821, this fixes the build issues, and additionally fixes the Wayland configuration as descirbed in that PR.

Now, since there hasn't been a response to https://github.com/rust-windowing/winit/commit/639e1d40b13a38d9545e452d9f7e7ae5b77e4740#r108100537, the question remains, should we fix that locally in our fork, or try to reach out to the winit guys once more.

I would really like to see this moving forward, not only for the reasons I described in my earlier message, but also because switching to a newer winit should close a lot of the bugs that we have currently.

File: src/window/renderer.rs
Comment (fredizzimo) [inline src/window/renderer.rs:35]: I think we should force a minimum size here, rather than crashing with unwrap errors. I will fix that myself.

Actually I will also add a maximum size of 8k or something, if someone has a bigger window than that, it's better to get a blurry one, than to crash because of too big buffers. I think we have some reports for that as well. I will add the same limit to the creation as well.

**Review by fredizzimo — CHANGES_REQUESTED** [review]

**Comment by shaunsingh** [comment]
> Now, since there hasn't been a response to https://github.com/rust-windowing/winit/commit/639e1d40b13a38d9545e452d9f7e7ae5b77e4740#r108100537, the question remains, should we fix that locally in our fork, or try to reach out to the winit guys once more.

Feel free to add it into the fork, I feel like it's quite an urgent issue, and can take quite some time to get patches merged upstream

File: src/renderer/opengl.rs
Comment (fredizzimo) [reply src/renderer/opengl.rs:88]: Only the vsync option was missing, so I added that.

File: src/window/renderer.rs
Comment (fredizzimo) [reply src/window/renderer.rs]: I added validation for all the buffer sizes.

**Comment by fredizzimo** [comment]
The keyboard reperat fix should now be included

**Review by fredizzimo — APPROVED** [review]

**Review by last-partizan — APPROVED** [review]
I've been using this branch for a few days, and it works great.

Anything else needs to be done before merging?

**Comment by fredizzimo** [comment]
Nothing from my part, except that perhaps it would make sense to lock the commit id of the new-keyboard-v3 branch in `Cargo.toml` instead of relying on just the lockfile to point to the right commit. That we can better control when to update to newer versions.

**Comment by MultisampledNight** [comment]
Sidenote that building this branch fails with `cargo install` or after `cargo update` is ran, with errors in winit and this warning:

```
warning: Patch `winit v0.28.1 (https://github.com/neovide/winit?branch=new-keyboard-v3#4d8d82fd)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
```

I've pushed a commit locking `winit` to `=0.28.1` which fixes this. Seemingly the implicit `^` specifier was not enough and upgraded to `0.28.6` unconditionally (and also fixes a small formatting issue), since this **was** `cargo tree -i winit` after `cargo update` is ran:

```
winit v0.28.1 (https://github.com/neovide/winit?branch=new-keyboard-v3#4d8d82fd)
├── glutin-winit v0.3.0
│   └── neovide v0.10.4 (/home/multisn8/info/lurk/neovide)
└── neovide v0.10.4 (/home/multisn8/info/lurk/neovide)
```

**Comment by fredizzimo** [comment]
We can't merge this. There seems to be a deadlock in the new keyboard handling of winit on Windows. I'm not sure what's wrong, and I'm don't think I have time to debug it today. But we did have a deadlock earlier too that we fixed and notified the upstream about. They did not take our fix as I wrote it  and instead they made another fix by their own, so I wonder if they messed it up. But it could of course be some other deadlock as well.

The deadlock is 100% for me, when I use neovide on a windows remote desktop after I minimize the desktop and return.

**Comment by fredizzimo** [comment]
It's indeed the same bug, or at least very similar, the code calls `PeekMessage` inside the keyboard event handling which deadlocks due to re-entrancy. 

It's either a new regression in the keyboard branch after our previous merge or their fix does not solve the problem https://github.com/ArturKovacs/winit/pull/1

**Comment by fredizzimo** [comment]
It looks like we might be able to switch over to point to the official branch instead. The deadlock appears to be fixed https://github.com/rust-windowing/winit/pull/2662 and the Wayland keyboard repeat might also be fixed. 

I have some minor changes that moves this over to that instead of our fork, along with a few minior fixes due to very slight interface changes. Should I update this PR with those changes, our should we use a separate one for that?

I don't think it makes much sense to pull in the changes to our fork and have Neovide point to that. If we encountered bugs that needs to be fixed, we should send PR upstream instead. And if we lock the commit hash along with the branch it should be just as stable as having our own fork.

**Comment by last-partizan** [comment]
I agree with you, it's better to use official branch.

**Comment by MultisampledNight** [comment]
Tested this in combination with https://github.com/neovide/neovide/pull/1870 for the past 25 minutes on NixOS with Sway (Wayland) with an Intel iGPU, and I cannot observe any change in behavior so far, nice!

**Comment by fredizzimo** [comment]
@last-partizan, could you try that with the official Winit release and on of their examples?  The switch to the upstream version also changes the main Winit version from 0.28.1 to 0.28.6, so it's possible that there's been some regressions.

I also noticed now when checking the diffs that my change converts `cargo.toml` to Linux line endings. I was not supposed to do that, but I think it's a good change, I don't know why we store it with windows line endings, and my mergetool does not detect the line endings automatically so it's been very annoying to deal with that file.

**Comment by MultisampledNight** [comment]
Actually I just noticed there is _one_ thing: If [`g:neovide_hide_mouse_when_typing`](https://neovide.dev/configuration.html#hiding-the-mouse-when-typing) is set to `v:true`, moving the mouse inside the Neovide window does not make it visible again — it needs to be moved out of bounds first. I'm not sure if this should be fixed on our end, since https://github.com/neovide/neovide/blob/4b3c234a49d7db3edbc6ba821964805525f7a02a/src/window/mouse_manager.rs#L443 _should_ trigger the mouse to be visible again, or if this is in winit.

(The file in question has not been changed here or in #1870.)

**Comment by fredizzimo** [comment]
@MultisampledNight, there's an example in winit that you could try `cargo run --example cursor_grab`. Pressing `h` should hide the cursor and `shift-h` show it.

**Comment by MultisampledNight** [comment]
@fredizzimo The winit examples don't display under Wayland, since they don't render. I'll create a wgpu test app and try it with that sometime today.

**Comment by MultisampledNight** [comment]
> Actually I just noticed there is _one_ thing: If [`g:neovide_hide_mouse_when_typing`](https://neovide.dev/configuration.html#hiding-the-mouse-when-typing) is set to `v:true`, moving the mouse inside the Neovide window does not make it visible again — it needs to be moved out of bounds first. I'm not sure if this should be fixed on our end, since
> 
> https://github.com/neovide/neovide/blob/4b3c234a49d7db3edbc6ba821964805525f7a02a/src/window/mouse_manager.rs#L443
> _should_ trigger the mouse to be visible again, or if this is due to winit.
> 
> (The file in question has not been changed here or in #1870.)

Fwiw this was fixed in rust-windowing/winit#2821, but that landed on winit's `master` branch, so we won't see it until https://github.com/rust-windowing/winit/pull/2662 is merged and we depend directly on winit's default branch/a new release after that instead.

**Comment by fredizzimo** [comment]
@last-partizan, I checked Wayland with `WINIT_UNIX_BACKEND=x11` 

Using KDE with Nvidia 3070 it does not crash for me, but flickers badly when resizing. Native Wayland works fine though. 

Did you have any chance to debug where it crashes yet? It would be nice to support xwayland, but if native wayland works, I don't think there's a need for it.

**Comment by last-partizan** [comment]
I does not crashes, but freezes, so i kill it with ctrl-c.

Is there any way to see on what line of code it freezed?

Here's how it looks like 
[Screencast from 2023-05-27 11-04-44.webm](https://github.com/neovide/neovide/assets/301015/6c938c54-15e4-4630-a718-dc2b8f8bd65b)

Here's last debug lines:
```
TRACE [neovide::channel_utils] neovide::window::WindowCommand SetMouseEnabled(true)
TRACE [neovide::editor] Image flushed
TRACE [neovide::channel_utils] alloc::vec::Vec<neovide::renderer::DrawCommand> [ModeChanged(Normal), UpdateCursor(Cursor { grid_position: (6, 1), parent_window_id: 1, shape: Block, cell_percentage: Some(0.0), blinkwait: Some(0), blinkon: Some(0), blinkoff: Some(0), style: None, enabled: true, double_width: false, grid_cell: (" ", None) })]
TRACE [neovide::redraw_scheduler] Next frame queued
```

And when i try to strace it, it looks like this:
```
strace: Process 9602 attached
futex(0x55a54339b228, FUTEX_WAIT_BITSET, 2, NULL, FUTEX_BITSET_MATCH_ANY) = ?
+++ killed by SIGKILL +++
```

Any ideas how to proceed?

Good news: I cannot reproduce it on my laptop, so hopefully not everyone is affected.

**Comment by fredizzimo** [comment]
@last-partizan 

Using a debugger is the easiest, just break into the debugger when the hang happens, and check which mutex/futex the main loop is blocking on. Neovide has many threads so you might need to expand the callstacks one by one to find out. If you need more help, you can post the full callstack here.

I'm usually using codelldb extension for Visual Studio code for debugging, since that works great for rust https://github.com/vadimcn/codelldb with the following launch configuration, the one we have in the repository is outdated

 ```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "lldb",
            "request": "launch",
            "args": ["--multigrid"],
            "cargo": {
                "args": ["build"]
            }
        },
    ]
}
```
That enables both debugging and building from within VSCode. I still write the code in Neovim/Neovide though, so VSCode is just for debugging. 

But of course you can also use lldb directly, or use it through Neovim DAP, but I just think that the Visual Studio debugging experience is better.

**Comment by MultisampledNight** [comment]
Fwiw another short and hakcy possibility would be to `kill -SEGV $(pgrep neovide)` and using `systemctl debug`, then in the launched gdb instance `backtrace` to display the backtrace it was hanging on.

**Comment by last-partizan** [comment]
I was able to start debugging session with lldb and codelldb, but i haven't figured out why it hangs.

<details><summary>Here's backtrace from lldb</summary>

```lldb
(lldb) bt
* thread #1, name = 'neovide', stop reason = signal SIGSTOP
  * frame #0: 0x00007f62f24072ed libc.so.6`syscall + 29
    frame #1: 0x00007f62e0ec71de radeonsi_dri.so`___lldb_unnamed_symbol2790 + 142
    frame #2: 0x00007f62e0f269fb radeonsi_dri.so`___lldb_unnamed_symbol4325 + 203
    frame #3: 0x00007f62e0eba210 radeonsi_dri.so`___lldb_unnamed_symbol2634 + 80
    frame #4: 0x00007f62e309c73f libGLX_mesa.so.0`___lldb_unnamed_symbol2575 + 79
    frame #5: 0x00007f62e30885bb libGLX_mesa.so.0`___lldb_unnamed_symbol2294 + 91
    frame #6: 0x00005596969770fd neovide`glutin_glx_sys::glx::Glx::SwapBuffers::h4b5b29833b4e6c72(self=0x00005596986730a0, dpy=0x0000559699af8da0, drawable=12582915) at glx_bindings.rs:561:120
    frame #7: 0x00005596968a5bcb neovide`_$LT$glutin..api..glx..surface..Surface$LT$T$GT$$u20$as$u20$glutin..surface..GlSurface$LT$T$GT$$GT$::swap_buffers::_$u7b$$u7b$closure$u7d$$u7d$::heccc6a8afa9764cd at surface.rs:215:13
    frame #8: 0x00005596968c45d9 neovide`glutin::api::glx::last_glx_error::hb9aa29a8c9e7cfff(display=(__0 = 0x0000559699af8da0), callback={closure_env#0}<glutin::surface::WindowSurface> @ 0x00007ffd24723858) at mod.rs:193:18
    frame #9: 0x00005596968a5b5f neovide`_$LT$glutin..api..glx..surface..Surface$LT$T$GT$$u20$as$u20$glutin..surface..GlSurface$LT$T$GT$$GT$::swap_buffers::h42505733f8afb379(self=0x00007ffd24731240, _context=0x00007ffd24731268) at surface.rs:214:9
    frame #10: 0x00005596969e1859 neovide`_$LT$glutin..surface..Surface$LT$T$GT$$u20$as$u20$glutin..surface..GlSurface$LT$T$GT$$GT$::swap_buffers::h7b1d5dd0cf6ed834(self=0x00007ffd24731238, context=0x00007ffd24731260) at surface.rs:314:17
    frame #11: 0x0000559696a147ea neovide`neovide::renderer::opengl::Context::swap_buffers::h2824e74af5de5207(self=0x00007ffd24731158) at opengl.rs:45:9
    frame #12: 0x000055969693dfe5 neovide`neovide::window::WinitWindowWrapper::draw_frame::hbaf693d67cfc7a2e(self=0x00007ffd24731158, dt=0.0227467846) at mod.rs:229:13
    frame #13: 0x00005596969b5be8 neovide`neovide::window::create_window::_$u7b$$u7b$closure$u7d$$u7d$::h72b767cade56deb1(e=Event<()> @ 0x00007ffd24723ef8, _window_target=0x0000559699bc97c0, control_flow=0x00007ffd2472c520) at mod.rs:518:13
    frame #14: 0x0000559696a9e042 neovide`winit::platform_impl::platform::x11::EventLoop$LT$T$GT$::drain_events::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::hbbabc9ebd002abe5(event=Event<()> @ 0x00007ffd24724158, window_target=0x0000559699bc97c0, control_flow=0x00007ffd2472c520) at mod.rs:553:29
    frame #15: 0x00005596969b4a85 neovide`winit::platform_impl::platform::sticky_exit_callback::h2d5178cbc5bbab45(evt=Event<()> @ 0x00007ffd24724238, target=0x0000559699bc97c0, control_flow=0x00007ffd2472c520, callback=0x00007ffd247242d0) at mod.rs:924:9
    frame #16: 0x0000559696a9dee3 neovide`winit::platform_impl::platform::x11::EventLoop$LT$T$GT$::drain_events::_$u7b$$u7b$closure$u7d$$u7d$::h4a908025594df04b(event=<unavailable>) at mod.rs:544:17
    frame #17: 0x000055969692ee18 neovide`winit::platform_impl::platform::x11::event_processor::EventProcessor$LT$T$GT$::process_event::hf48cc20b24926f68(self=0x00007ffd2472c8a0, xev=0x00007ffd2472baa0, callback={closure_env#0}<(), neovide::window::create_window::{closure_env#1}> @ 0x00007ffd2472bc20) at event_processor.rs:450:25
    frame #18: 0x0000559696a9de15 neovide`winit::platform_impl::platform::x11::EventLoop$LT$T$GT$::drain_events::hb988dedc72d2c6d4(self=0x00007ffd2472c838, callback=0x00007ffd24731148, control_flow=0x00007ffd2472c520) at mod.rs:543:13
    frame #19: 0x0000559696a9cdff neovide`winit::platform_impl::platform::x11::EventLoop$LT$T$GT$::run_return::single_iteration::h594fc8265cd9b880(this=0x00007ffd2472c838, control_flow=0x00007ffd2472c520, cause=0x00007ffd2472c548, callback=0x00007ffd24731148) at mod.rs:374:13
    frame #20: 0x0000559696a9db1b neovide`winit::platform_impl::platform::x11::EventLoop$LT$T$GT$::run_return::h44f7d6741b033d5a(self=0x00007ffd2472c838, callback={closure_env#1} @ 0x00007ffd24731148) at mod.rs:514:27
    frame #21: 0x0000559696a9e16d neovide`winit::platform_impl::platform::x11::EventLoop$LT$T$GT$::run::h292483a29ba25803(self=<unavailable>, callback=<unavailable>) at mod.rs:529:25
    frame #22: 0x00005596969b4912 neovide`winit::platform_impl::platform::EventLoop$LT$T$GT$::run::h443392cdf7033a9a(self=EventLoop<()> @ 0x00007ffd2472cf08, callback={closure_env#1} @ 0x00007ffd24731148) at mod.rs:832:56
    frame #23: 0x0000559696909bd9 neovide`winit::event_loop::EventLoop$LT$T$GT$::run::h45bbc2de0c89b18b(self=<unavailable>, event_handler={closure_env#1} @ 0x00007ffd24731148) at event_loop.rs:305:9
    frame #24: 0x0000559696940105 neovide`neovide::window::create_window::hd550cbd955f3d46e at mod.rs:474:5
    frame #25: 0x00005596968914ed neovide`neovide::protected_main::hfbb130847588a2ed at main.rs:168:5
    frame #26: 0x00005596968913b8 neovide`neovide::main::hfbbcab7472396a4c at main.rs:73:5
    frame #27: 0x0000559696a5d58b neovide`core::ops::function::FnOnce::call_once::ha6e66dcff1d39b27((null)=(neovide`neovide::main::hfbbcab7472396a4c at main.rs:63), (null)=<unavailable>) at function.rs:250:5
    frame #28: 0x000055969698369e neovide`std::sys_common::backtrace::__rust_begin_short_backtrace::ha13f6e4d92f57739(f=(neovide`neovide::main::hfbbcab7472396a4c at main.rs:63)) at backtrace.rs:134:18
    frame #29: 0x0000559696a10291 neovide`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::hc350b52dc7d41547 at rt.rs:166:18
    frame #30: 0x00005596978d992c neovide`std::rt::lang_start_internal::h2947494479068ade [inlined] core::ops::function::impls::_$LT$impl$u20$core..ops..function..FnOnce$LT$A$GT$$u20$for$u20$$RF$F$GT$::call_once::h814f0ad687ab2f92(self=&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe) @ 0x0000557dbe49c150, args=<unavailable>) at function.rs:287:13
    frame #31: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade at panicking.rs:487:40
    frame #32: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade at panicking.rs:451:19
    frame #33: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade [inlined] std::panic::catch_unwind::he1669253985d4352(f=&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe) @ 0x0000557dbe49c2f0) at panic.rs:140:14
    frame #34: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade [inlined] std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::hd778e2099ecb758e at rt.rs:148:48
    frame #35: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade at panicking.rs:487:40
    frame #36: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade at panicking.rs:451:19
    frame #37: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade [inlined] std::panic::catch_unwind::hb9e5b8ae73c93b8f(f={closure_env#2} @ 0x0000557dbe49c2f0) at panic.rs:140:14
    frame #38: 0x00005596978d9929 neovide`std::rt::lang_start_internal::h2947494479068ade(main=&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe) @ 0x0000557dbe49c2f0, argc=<unavailable>, argv=<unavailable>, sigpipe=<unavailable>) at rt.rs:148:20
    frame #39: 0x0000559696a1026a neovide`std::rt::lang_start::h5c51dd532b36ed19(main=(neovide`neovide::main::hfbbcab7472396a4c at main.rs:63), argc=1, argv=0x00007ffd24731f88, sigpipe='\0') at rt.rs:165:17
    frame #40: 0x0000559696892b7e neovide`main + 30
    frame #41: 0x00007f62f2327850 libc.so.6`___lldb_unnamed_symbol3141 + 128
    frame #42: 0x00007f62f232790a libc.so.6`__libc_start_main + 138
    frame #43: 0x000055969687bd75 neovide`_start + 37
```
</details>

And screenshots from vscode:
image


Looking at 
```
    frame #1: 0x00007f62e0ec71de radeonsi_dri.so`___lldb_unnamed_symbol2790 + 142
```

 it could be anything driver-related.

We have workaround, so feel free to ignore this issue.

**Comment by fredizzimo** [comment]
@last-partizan,

I found a few of tangential issues https://github.com/rust-windowing/glutin/issues/1231, https://gitlab.freedesktop.org/mesa/mesa/-/issues/4932 and https://gitlab.freedesktop.org/mesa/mesa/-/issues/4180, but none of those matches exactly.

But it does seems like setting a swap interval (vsync) is a bad idea on Wayland. I bet `--novsync` also "fixes" the issue at the loss of some smoothness. My smooth rendering thread PR for example will switch to a software timer in that case.

Fortunately there's a better way to it in Wayland using https://wayland-book.com/surfaces-in-depth/frame-callbacks.html. Especially if `--novsync` works for you I will implement that in my smooth rendering PR. It will also fix the issue that it does not support a `g:neovide_refresh_rate` lower than the actual compositor refresh rate. So for people not  manually increasing `g:neovide_refresh_rate` from 60 to 120 for example, it will also be smoother by default.

**Comment by last-partizan** [comment]
`--novsync` has no effect on this freeze.

**Comment by fredizzimo** [comment]
@last-partizan, 

Hm, if `--novsync` doesn't have any effect, I wonder if it really blocks on `swap_buffers`? 

If you break and there's no hang, it will very likely break there, since most of the rendering loop will be waiting for vsync.

Could you try again, and ensure that neovide is really hung when you break it in the debugger? Maybe using `--novsync` this time.


BTW, meanwhile I have almost implemented vsync using frame callbacks, but now it seems like I have a problem with the event dispatching that I don't really know how to solve at the moment.

**Comment by fredizzimo** [comment]
Unfortunately it looks like using frame callbacks is a dead end. Well it works, but it won't allow us to draw every second frame for example. In order to do that `wl_surface.commit()` has to be called, and that's done deep inside winit, actually I think in `stck-adwaita`. So I don't think it's a good idea to make any extra calls, and we probably also need to actually draw something when doing that, since the buffers will be swapped.

Anyway, upon further research, this is exactly what `SwapBuffers` does internally anyway, but our own implementation would allow a timeout, and therefore not block forever. So in theory it could fix the hang, but if disabling `vsync` doesn't do it, then it won't work anyway.

But on the more positive note, the upsream branch has been merged into winit master https://github.com/rust-windowing/winit/pull/2662 has been merged now while writing this message. I will update this PR to follow that commit instead.

**Comment by fredizzimo** [comment]
We are now finally tracking winit/master! 

There were some quite major refactoring in https://github.com/rust-windowing/winit/pull/2817, since the the last keyboard update, which mostly affects the modifiers. But based on my tests the modifiers appears to work at least on Wayland, which I'm currently on, so no regressions there. But more testing is needed on other platforms before merging.

The biggest concern is probably the Mac modifiers with it's special handling of Option/Alt.

**Comment by last-partizan** [comment]
I tried running debugger with `--novsync`, and it hangs just like before.

image

**Comment by fredizzimo** [comment]
@last-partizan, have you tried the https://github.com/neovide/neovide/pull/1870 branch, it has a separate thread for rendering. I don't think it fixes anything, but worth a shot. 

Anyway, I don't think it's important to fix the XWayland support, since native Wayland works just fine, at most it will cause hangs for users, that used it as a workaround before.

In order to debug it further, we would need symbols for the the glx implementation. It's possible to debug further, but unless we really have to I rather spend my time on something more important.

**Comment by last-partizan** [comment]
@fredizzimo yes, that branch fixes freezing.

**Comment by fredizzimo** [comment]
So far I haven't found anything, either on Windows or Wayland that would prevent us from merging this.

Well, maybe the issue with needing `--nosrgb` at least on some systems. BTW, wezterm also has the same or a very similar issue https://github.com/wez/wezterm/issues/1701, so it's possible that there's something with winit/glutin.

**Comment by fredizzimo** [comment]
I found the problems with SRGB, but I don't have time to properly test and fix it today

This needs to use srgb when the framebuffer is SRGB

https://github.com/neovide/neovide/blob/f623704ffa17d0ba6388d9db01994ac448238175/src/window/renderer.rs#L68-L70


And so does this
https://github.com/neovide/neovide/blob/f623704ffa17d0ba6388d9db01994ac448238175/src/window/renderer.rs#L32-L40

Those lines have not been changed by this PR, and I'm confused how this has actually ever worked, and why it works on other platforms.  Some more investigation is needed, otherwise the fix would be trivial.

**Comment by fredizzimo** [comment]
I have been reading up and experimenting a bit more with srgb stuff. And my conclusion is that at least on Wayland with Nvidia it does not work as it should.

Creating a backbuffer with `srgb` support should by itself not enable `srgb`, it just makes it possible to do so. From this
https://www.khronos.org/opengl/wiki/Image_Format
> [When images with this format are used as a render target](https://www.khronos.org/opengl/wiki/Framebuffer_Colorspace), OpenGL will automatically convert the output colours from linear to the sRGB colorspace if, and only if, GL_FRAMEBUFFER_SRGB is enabled. The alpha will be written as given. When writing multiple outputs, only outputs written to sRGB image formats will undergo such conversion.

`GL_FRAMEBUFFER_SRGB` is automatically enabled by Skia when we tell it to, but  we are not. We request linear formats, so there should be no conversion. To properly request srgb we should set `ColorType::SRGBA8888` and then either convert all our colours to srgb or pass `Some(ColorSpace::new_srgb_linear()),` That gives us the correct colours on the other platforms, but on Wayland with Nvidia we need to either pass `None` or `Some(ColorSpace::new_srgb())`. I think the automatic conversion from linear to srgb is not working, so the output ends up being too dark.

Anyway, since we are not really using srgb, I suggest that we disable it by default, except on Windows, where we need it to workaround this bug https://github.com/neovide/neovide/issues/907.

I also found out that the environment variable `NEOVIDE_NO_SRGB` also works wrong, `NEOVIDE_NO_SRGB=0` means `--nosrgb` while `NEOVIDE_NO_SRGB=1` means srgb enabled. It would be nice if the environment variable was named `NEOVIDE_SRGB` and that we had two command line options `--nosrgb` and `--srgb`.

If someone has time to do those modifications, feel free to do so. It's getting late here, so no more changes by me tonight.

**Comment by fredizzimo** [comment]
I created a separate PR for the srgb issues https://github.com/neovide/neovide/pull/1885, so we should preferably merge that one first.

**Comment by last-partizan** [comment]
@fredizzimo i want to convert `Cargo.toml` to unix line endings in the separate branch, and merge it first. But it will introduce conflicts in this branch. Is that okay? Or i can wait until this merged.

**Comment by fredizzimo** [comment]
@last-partizan, go ahead. I think I accidentally already converted it here.

Actually I would prefer if you also sorted it alphabetically, so that it’s easier to find exising dependencies. I think that I already did that in one of my other pull requests. That also slightly reduces the risk of merge conflicts in the future, since you don’t always add new dependencies  to the end.

**Comment by fredizzimo** [comment]
I have rebased this (and https://github.com/neovide/neovide/pull/1787, https://github.com/neovide/neovide/pull/1790, https://github.com/neovide/neovide/pull/1870) on top of the latest master.

**Comment by last-partizan** [comment]
@fredizzimo  Is there any blocking issues?

I have used #1870 branch (based on this) for a week now, and haven't noticed any keyboard-related issues. Linux/x11 and wayland works fine.

@Kethku Is there anything you want to check before merging this?

**Comment by fredizzimo** [comment]
@last-partizan. I don't know of any blockers either, unless we want to wait for the planned beta https://github.com/rust-windowing/winit/issues/2686. But I think it makes sense to merge now, and update to the beta as soon as it is out.

**Comment by shaunsingh** [comment]
LGTM, tested on macOS

For some reason, trying to open neovide from finder on macOS no longer works, but if I do `open -a /Applications/Neovide.app` and run it from the terminal ... it opens just fine. Not sure if that's a result of these changes or not

**Comment by last-partizan** [comment]
Let's merge this.

I was about mo merge it myself, but i think @fredizzimo can do better job at writing commit message.

@fredizzimo can you do it? And please, use "Squash and merge", and [conventianl commits](https://www.conventionalcommits.org/en/v1.0.0/) syntax when writing commit message.

**Comment by last-partizan** [comment]
I merged main into this, and resolved conflicts in the "Cargo.lock", by resetting it to "main" branch version and then running `cargo r`.

**Comment by fredizzimo** [comment]
Later today took a couple of days, but now it's merged :) Let's see what all this fixes and breaks.

**Comment by fredizzimo** [comment]
I quickly went through all our open issues to see what could potentially be fixed, but I probably missed a lot of issues, so it would be good to do another pass at some point.

It looks like the update itself might not have fixed that many keyboard issues, we still need to do some work especially around the modifier support.

**Comment by manfredlotz** [comment]
Once I had opened #1236.

My system: EndeavourOS using Xorg

Having build neovide from commit _4b3c234a49d7db3edbc6ba821964805525f7a02a_ (May 14) and neovide doesn't react when having started neovide and switched to other app and switching back to neovide. Only moving the mouse makes it work.

Now haven git pulled f412399676c83ea670cd22745b850dad9ee210a0 (June 14) and rebuilt neovide. 

Problem seems to be fixed as when switching to other apps and switching back to neovide then neovide is responsive without any intermediate mouse movement.

So, from my side things seems to be fixed! Thanks a lot for your work.

**Pull request merged**
